src/tools/schema2graph.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Schema} from '../runtime/schema.js';
+import {Direction} from '../runtime/manifest-ast-nodes.js';
+import {Dictionary} from '../runtime/hot.js';
+import {Manifest} from '../runtime/manifest.js';
+
+export interface FieldEntry {
+  particleName: string;
+  schema: Schema;
+  direction: Direction;
+  connectionName: string;
+}
+
+export interface TypeLattice {
+  nodes: Dictionary<FieldEntry>;
+  edges: Dictionary<string[]>;
+  nodesByParticle: Dictionary<FieldEntry[]>;
+  contains(name: string): boolean;
+}
+
+class Builder {
```

Comment:
I don't think the separation of Builder and TypeGraph is especially helpful here:
- The from/to/build API feels like it's exposing internals to the caller, especially since there's no real way (or reason to) use the class any other way.
- Having a builder suggests that TypeGraph is immutable, but ~half of its LOC are taken up by the construction methods `addNode` and `addEdge`.